pddl/logic/functions.py

```diff
@@ -0,0 +1,366 @@
+#
+# Copyright 2021-2023 WhiteMech
+#
+# ------------------------------
+#
+# This file is part of pddl.
+#
+# Use of this source code is governed by an MIT-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+#
+
+"""This class implements PDDL functions."""
+import functools
+from typing import Sequence
+
+from pddl.custom_types import namelike, parse_function
+from pddl.helpers.base import assert_
+from pddl.helpers.cache_hash import cache_hash
+from pddl.logic.base import Atomic, MonotoneOp
+from pddl.logic.terms import Term
+from pddl.parser.symbols import Symbols
+
+
+@cache_hash
+class FunctionExpression(Atomic):
+    """A class for all the function expressions."""
+
+
+@cache_hash
+@functools.total_ordering
```

Comment:
If I am not mistaken, we used `total_ordering` for any leaf expression in some expression hierarchy (in this case, function expression). `NumericFunction` is a leaf expression (correct?) so it is correctly annotated as with `total_ordering`.

What about the other _leaf_ expression below? `NumericValue`, `Metric` etc. If I recall correctly, tThe rationale is that: if instances of such types can appear in a sequence of some binary operator, then we should be able to sort them (if the operator is commutative). I guess this was useful in other places for canonicity.

Reply:
While `NumericValue` can appear in some binary operators, `Metric` does not. However, `Metric` is a leaf expression.